internal/discovery/discovery.go

```diff
@@ -0,0 +1,223 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package discovery provides a discovery and resolution logic for GVKs.
+package discovery
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/dynamic/dynamicinformer"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog/v2"
+
+	"k8s.io/kube-state-metrics/v2/internal/store"
+	"k8s.io/kube-state-metrics/v2/pkg/customresource"
+	"k8s.io/kube-state-metrics/v2/pkg/metricshandler"
+	"k8s.io/kube-state-metrics/v2/pkg/options"
+	"k8s.io/kube-state-metrics/v2/pkg/util"
+)
+
+// PopulateTimeout is the timeout on populating the cache for the first time.
+const PopulateTimeout = 10 * time.Second
+
+// Interval is the time interval between two cache sync checks.
+const Interval = 3 * time.Second
+
+// StartDiscovery starts the discovery process, fetching all the objects that can be listed from the apiserver, every `Interval` seconds.
+// resolveGVK needs to be called after StartDiscovery to generate factories.
+func (r *CRDiscoverer) StartDiscovery(ctx context.Context, config *rest.Config) error {
+	dClientset := dynamic.NewForConfigOrDie(config)
+	factory := dynamicinformer.NewFilteredDynamicInformer(dClientset, schema.GroupVersionResource{
+		Group:    "apiextensions.k8s.io",
+		Version:  "v1",
+		Resource: "customresourcedefinitions",
+	}, "", 0, nil, nil)
+	informer := factory.Informer()
+	stopper := make(chan struct{})
+	_, err := informer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			objSpec := obj.(*unstructured.Unstructured).Object["spec"].(map[string]interface{})
+			for _, version := range objSpec["versions"].([]interface{}) {
+				gotGVK := schema.GroupVersionKind{
+					Group:   objSpec["group"].(string),
+					Version: version.(map[string]interface{})["name"].(string),
+					Kind:    objSpec["names"].(map[string]interface{})["kind"].(string),
+				}
+				r.AppendToMap(gotGVK)
+				r.SafeWrite(func() {
+					r.WasUpdated = true
+				})
+			}
+		},
+		DeleteFunc: func(obj interface{}) {
+			objSpec := obj.(*unstructured.Unstructured).Object["spec"].(map[string]interface{})
+			for _, version := range objSpec["versions"].([]interface{}) {
+				gotGVK := schema.GroupVersionKind{
+					Group:   objSpec["group"].(string),
+					Version: version.(map[string]interface{})["name"].(string),
+					Kind:    objSpec["names"].(map[string]interface{})["kind"].(string),
+				}
+				r.RemoveFromMap(gotGVK)
+				r.SafeWrite(func() {
+					r.WasUpdated = true
+				})
+			}
+		},
+	})
+	if err != nil {
+		return err
+	}
+	// Respect context cancellation.
+	go func() {
+		for range ctx.Done() {
```

Comment:
I don't understand this loop (might be my lack of golang knowledge on channels). Can you explain?

Reply:
Thanks! Is this a busy loop and will run constantly or will it block waiting for ctx.Done() to return?

Reply:
It'll block, hence the goroutine. :)

Reply:
TIL! Thanks for the explanation!